crates/floresta-wire/src/p2p_wire/tests/chain_selector.rs

```diff
@@ -0,0 +1,537 @@
+    use super::tests_utils::get_essentials;
+    use super::tests_utils::setup_test;
+
+    #[async_std::test]
+    async fn accept_one_header() {
+        let (headers, _, _, _) = get_essentials();
+
+        setup_test(
+            "test_chain_selector",
+            vec![(headers, HashMap::new(), HashMap::new())],
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn two_peers_different_tips() {
+        let (mut headers, _, _, _) = get_essentials();
+
+        let mut peers = Vec::new();
+
+        for _ in 0..2 {
+            headers.pop();
+            peers.push((headers.clone(), HashMap::new(), HashMap::new()))
+        }
+
+        setup_test(
+            "two_peers_different_tips",
+            peers,
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn ten_peers_different_tips() {
+        let (mut headers, _, _, _) = get_essentials();
+
+        let mut peers = Vec::new();
+
+        for _ in 0..10 {
+            headers.pop();
+            headers.pop();
+
+            peers.push((headers.clone(), HashMap::new(), HashMap::new()))
+        }
+
+        setup_test(
+            "ten_peers_different_tips",
+            peers,
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn two_peers_one_lying() {
+        let (mut headers, blocks, true_filters, tip_hash) = get_essentials();
+        headers.truncate(120);
+
+        // Create a random false utreexo-acc for the lying peer
+        let mut false_filters = true_filters.clone();
+        false_filters.remove(&tip_hash).unwrap();
+        false_filters.insert(tip_hash, create_false_acc(119));
+
+        let peers = vec![
+            (headers.clone(), blocks.clone(), true_filters),
+            (headers, blocks, false_filters),
+        ];
+
+        setup_test(
+            "two_peers_one_lying",
+            peers,
+            true,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+
+    #[async_std::test]
+    async fn ten_peers_one_honest() {
+        let (mut headers, blocks, true_filters, tip_hash) = get_essentials();
+        headers.truncate(120);
+
+        let mut false_filters = true_filters.clone();
+        let mut peers = Vec::new();
+
+        // Create 9 lying peers
+        for _ in 0..9 {
+            false_filters.remove(&tip_hash).unwrap();
+            false_filters.insert(tip_hash, create_false_acc(119));
+            peers.push((headers.clone(), blocks.clone(), false_filters.clone()));
+        }
+        peers.push((headers, blocks, true_filters));
+
+        setup_test(
+            "ten_peers_one_honest",
+            peers,
+            true,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+    }
+}
```

Comment:
Could you add cases where a block is actually invalid?

Reply:
Please correct if I am going wrong here, When the chain selector node is run, it first achieves the best block headers from all its peers. Then if pow_fraud_proofs is enabled the `check_tips` function is executed, and all the peers have to give their best utreexo states, and if they differ then someone is lying.
 
The only time the whole UtreexoBlock data is used in `.update_acc` function where I think we calculate the immediate next utreexo acc on our own codebase. 

I couldn't find any code in `chain_selector.rs` where the block is actually validated according to the consensus rules. 

Maybe the tests for validity of blocks could be done in the `running_node` or the `sync_node`. I am a little confused here.

Reply:
I was thinking, like 3 different testing environments could be made, just like for `chain_selector` is done here, for `running_node` and `sync_node` and covering all their functionalities in their respective test codebases. 
So I think each and every function in `chain_selector.rs` is tested along, except `.is_our_chain_invalid` because I could not figure out how to create different tips in our chain, I need a little help there. 
After this, maybe I could move forward with writing tests for the other two remaining node ?? @Davidson-Souza your thoughts ?

Reply:
[This function](https://github.com/Davidson-Souza/Floresta/blob/5e96214fca6618f0643f13c678f2d4a5801154c8/crates/floresta-wire/src/p2p_wire/chain_selector.rs#L505) will check if our current best chain has an invalid block. 

> I was thinking, like 3 different testing environments could be made, just like for chain_selector is done here, for running_node and sync_node and covering all their functionalities in their respective test codebases.

Agree

> because I could not figure out how to create different tips in our chain, I need a little help there.

By having peers following different tips. They need to be forks, no only lag a few blocks. When `poke_peers` is called, it'll download the other tips.